frontend/test/metabase-lib/lib/Question.unit.spec.js

```diff
@@ -459,34 +458,67 @@ describe("Question", () => {
         const question = orders_count_question
           .setDisplay("funnel")
           .lockDisplay()
-          .maybeUnlockDisplay(["table", "scalar"])
-          .setDefaultDisplay();
+          .maybeResetDisplay(["table", "scalar"]);
 
         expect(question.display()).toBe("scalar");
       });
     });
 
-    describe("maybeUnlockDisplay", () => {
-      it("should keep display locked when it was locked with nonsense display", () => {
+    describe("maybeResetDisplay", () => {
+      it("should do nothing when it was locked with nonsense display", () => {
         const sensibleDisplays = ["table", "scalar"];
         const previousSensibleDisplays = sensibleDisplays;
         const question = new Question(orders_count_card, metadata)
           .setDisplay("funnel")
           .lockDisplay()
-          .maybeUnlockDisplay(sensibleDisplays, previousSensibleDisplays);
+          .maybeResetDisplay(sensibleDisplays, previousSensibleDisplays);
 
         expect(question.displayIsLocked()).toBe(true);
+        expect(question.display()).toBe("funnel");
       });
 
-      it("should unlock display it was locked with sensible display which has become unsensible", () => {
+      it("should use default display when nonsense display is used and was not locked", () => {
+        const sensibleDisplays = ["table", "scalar"];
+        const question = base_question
+          .setDisplay("funnel")
+          .maybeResetDisplay(sensibleDisplays, sensibleDisplays);
+
+        expect(question.display()).not.toBe("funnel");
+        expect(question.display()).toBe("table");
+      });
+
+      it("should unlock and use new sensible display when it was locked with sensible display which has become not sensible", () => {
         const previousSensibleDisplays = ["funnel"];
         const sensibleDisplays = ["table", "scalar"];
         const question = orders_count_question
           .setDisplay("funnel")
           .lockDisplay()
-          .maybeUnlockDisplay(sensibleDisplays, previousSensibleDisplays);
+          .maybeResetDisplay(sensibleDisplays, previousSensibleDisplays);
 
         expect(question.displayIsLocked()).toBe(false);
+        expect(question.display()).not.toBe("funnel");
+        expect(sensibleDisplays).toContain(question.display());
+      });
+
+      it("should keep any sensible display when display was locked", () => {
+        const sensibleDisplays = ["table", "scalar"];
+        const question = base_question
+          .setDisplay("scalar")
+          .lockDisplay()
+          .maybeResetDisplay(sensibleDisplays);
+
+        expect(question.display()).not.toBe("table");
+        expect(question.display()).toBe("scalar");
+      });
+
+      it("should keep any sensible display when display was not locked (metabase#30057)", () => {
```

Comment:
I'm not sure I understand this test case. This emphasizes the condition "when display was not locked". But the test would still pass when we call `.lockDisplay()` on the question.

I think this test exercises this code.
https://github.com/metabase/metabase/blob/f3583fae60d5281f9bcbaed18046fd8a4397ad66/frontend/src/metabase-lib/Question.ts#L322-L325

maybe the test name should be something along the line "should keep any sensible display when the default display is table"?

Reply:
> But the test would still pass when we call `.lockDisplay()` on the question.

The test case above this one is exactly for that particular case.

> This emphasizes the condition "when display was not locked"

I wanted to cover the case "regardless of whether display is locked", but since this requires 2 different setups, I split it into 2 test cases.